src/write_kosak.c

```diff
@@ -0,0 +1,2159 @@
+    return NULL;
+  }
+  return result;
+}
+
+static void wg_credential_ctx_destroy(credential_ctx_t *ctx) {
+  if (ctx == NULL) {
+    return;
+  }
+  if (ctx->private_key != NULL) {
+    EVP_PKEY_free(ctx->private_key);
+  }
+  sfree(ctx->email);
+  sfree(ctx);
+}
+
+static EVP_PKEY *wg_credential_contex_load_pkey(char const *filename,
+                                                char const *passphrase) {
+  OpenSSL_add_all_algorithms();
+  FILE *fp = fopen(filename, "rb");
+  if (fp == NULL) {
+    ERROR("write_kosak: Failed to open private key file %s", filename);
+    return NULL;
+  }
+
+  PKCS12 *p12 = d2i_PKCS12_fp(fp, NULL);
+  fclose(fp);
+  char err_buf[1024];
+  if (p12 == NULL) {
+    ERR_error_string_n(ERR_get_error(), err_buf, sizeof (err_buf));
+    ERROR("write_kosak: Reading private key %s failed: %s", filename, err_buf);
+    return NULL;
+  }
+
+  EVP_PKEY *pkey = NULL;
+  X509 *cert = NULL;
+  STACK_OF(X509) *ca = NULL;
+  int result = PKCS12_parse(p12, passphrase, &pkey, &cert, &ca); // 0 is failure
+  if (result == 0) {
+    ERR_error_string_n(ERR_get_error(), err_buf, sizeof (err_buf));
+    ERROR("write_kosak: Parsing private key %s failed: %s", filename, err_buf);
+    PKCS12_free(p12);
+    return NULL;
+  }
+
+  sk_X509_pop_free(ca, X509_free);
+  X509_free(cert);
+  PKCS12_free(p12);
+  return pkey;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// CURL submodule.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+
+// Does an HTTP GET or POST, with up to two optional HTTP headers. If body is
+// NULL, does a GET, otherwise does a POST.
+static int wg_curl_get_or_post(
+    char *response_buffer, size_t response_buffer_size,
+    const char *url, const char *header0, const char *header1,
+    const char *body);
+
+//------------------------------------------------------------------------------
+// Private implementation starts here.
+//------------------------------------------------------------------------------
+typedef struct {
+  char *data;
+  size_t size;
+} wg_curl_write_ctx_t;
+
+static size_t wg_curl_write_callback(char *ptr, size_t size, size_t nmemb,
+                                     void *userdata);
+
+static int wg_curl_get_or_post(
+    char *response_buffer, size_t response_buffer_size,
+    const char *url, const char *header0, const char *header1,
+    const char *body) {
+  const char *get_or_post_tag = body == NULL ? "GET" : "POST";
+  ERROR("Doing: %s request: url %s, header0 %s, header1 %s, body %s",
+        get_or_post_tag, url, header0, header1, body);
+  CURL *curl = curl_easy_init();
```

Comment:
Need to call this in the plugin init or it leads to crashes later.  See 66b400ab01b8133e450bb002e175117a1ab6f9ae